check.sh

```diff
@@ -8,139 +8,195 @@
+        if [[ "$godotVersion" =~ ^4\. ]]; then
+            log "Found 'godot' executable with version $godotVersion"
             godotBin="godot"
         else
-            echo "Found 'godot' executable, but it has the incompatible version $(godot --version)"
-            exit 2
+            log "Found 'godot' executable, but it has incompatible version $godotVersion"
+            return 1
         fi
 
-    # Error case
+    # Error case.
     else
-        echo "Godot executable not found"
-        exit 2
+        log "Godot executable not found; try setting GODOT4_BIN to the full path to the executable"
+        return 1
     fi
 }
 
+################################################################################
+# Commands
+################################################################################
+
+# Poor human's namespacing: all commands are prefixed with `cmd_` to avoid confusion with shell
+# builtins like `test`.
+
+function cmd_fmt() {
+    run cargo fmt --all -- --check
+}
+
+function cmd_clippy() {
+    run cargo clippy "${extraCargoArgs[@]}" -- \
+        -D clippy::suspicious \
+        -D clippy::style \
+        -D clippy::complexity \
+        -D clippy::perf \
+        -D clippy::dbg_macro \
+        -D clippy::todo \
+        -D clippy::unimplemented \
+        -D warnings
+}
+
+function cmd_test() {
+    run cargo test "${extraCargoArgs[@]}"
+}
+
+function cmd_itest() {
+    findGodot && \
+        run cargo build -p itest "${extraCargoArgs[@]}" && \
+        run "$godotBin" --path itest/godot --headless
+}
+
+function cmd_doc() {
+    run cargo doc --lib -p godot --no-deps "${extraCargoArgs[@]}"
+}
+
+function cmd_dok() {
+    run cargo doc --lib -p godot --no-deps "${extraCargoArgs[@]}" --open
+}
+
+################################################################################
+# Argument parsing
+################################################################################
+
 cmds=()
-extraArgs="${extraArgs[@]}"
+extraCargoArgs=()
 
-for arg in "${args[@]}"; do
+for arg in "$@"; do
     case "$arg" in
-    fmt)
-        cmds+=("cargo $toolchain fmt --all -- --check")
-        ;;
-    clippy)
-        cmds+=("cargo $toolchain clippy $extraArgs -- -D clippy::suspicious -D clippy::style -D clippy::complexity -D clippy::perf -D clippy::dbg_macro -D clippy::todo -D clippy::unimplemented -D warnings")
-        ;;
-    test)
-        cmds+=("cargo $toolchain test $extraArgs")
-        ;;
-    itest)
-        findGodot
-
-        cmds+=("cargo $toolchain build -p itest $extraArgs")
-        cmds+=("$godotBin --path itest/godot --headless")
-        ;;
-    doc)
-        cmds+=("cargo $toolchain doc --lib -p godot --no-deps $extraArgs")
-        ;;
-    dok)
-        cmds+=("cargo $toolchain doc --lib -p godot --no-deps $extraArgs --open")
-        ;;
-    *)
-        echo "Unrecognized command '$arg'"
-        exit 2
-        ;;
+        -h | --help | help)
+            echo "$HELP_TEXT"
+            exit 0
```

Comment:
Probably `-h|--help` should be enough, no? Just to streamline usage a bit.

Reply:
Plain `help` without any hyphens is a hidden feature, super simple to add and might save someone a few seconds if they try it as a guess. Documenting it would just clutter help output, since we'd also need to specify that it's not like other commands.

Which reminds me, let me point at `--help` in the "Unrecognized argument" error message.